### PR TITLE
sort edges during propagation to skip more edges

### PR DESCRIPTION
**Pull request by MaxOstrowski**
PR to talk about idea.
Sorting the edges that are added successively during one propagate call.
Idea:
Negative and small weights on the edge are more likely to produce:
1. negative cycles
2. changes of potential in the graph

By first adding edges with a lower weight, edges with smaller weights may be skipped afterwards.
What do you think? (First benchmarks on a small instance set show now difference)

### CODE REVIEW

**Comment by MaxOstrowski** [comment]
Remark: Difference in propagation speed is hard to measure as order of added clauses (and therefore the search) changes. The ratio between added and skipped edges could maybe used ?

**Comment by rkaminsk** [comment]
If you think that this might help, you could benchmark it. It would also be a good idea to make such a benchmark set available to everyone.

Otherwise, I think it is the right time to completely rewrite clingo-dl. We have learned that full propagation does not help at all but just leads to very complicated code. Adding something like clingcon's bound propagation is much easier to implement.

You should take a little more time with the code you are committing:
- only add edges if they are active
- to avoid allocations, edges could be a member variable
- best make sorting an option right away so that benchmarking this gets easier
- I do not think that the cost of sorting will be a problem as the dijsktra algorithm is more expensive anyway.

**Comment by MaxOstrowski** [comment]
> You should take a little more time with the code you are committing

Sure, was not meant to be merged, i just wanted to sketch the idea.
Do you think this could be the correct heuristic ? Should we consider the current potential of the nodes ? Should we resort after each added edge ? Does the "<" make sense at all ?



Benchmarks are running and looking promising on the xxx_shop benchmarks.
Have to rerun them as this run also included removing the satisfied clauses.

**Comment by MaxOstrowski** [comment]
[bench.zip](https://github.com/potassco/clingoDL/files/4250511/bench.zip)

Added an ods file.
clingodl-1 is the conda version, clingodl-3 is the wip which does not add satisfied clauses, and clingodl-2 is additionally sorting the edges.
Not adding satisfied clauses actually has quite an effect (for what it does, or not does). Sorting additional has some influence and luck with some instances.

**Comment by rkaminsk** [comment]
> Sure, was not meant to be merged, i just wanted to sketch the idea.

But if it is so easy to do it properly then just do it right away. This saves so much noise.

> Do you think this could be the correct heuristic ? Should we consider the current potential of the nodes ? Should we resort after each added edge ? Does the "<" make sense at all ?

I do not know. It is a heuristic after all. I would just benchmark it and keep what works best.

**Comment by rkaminsk** [comment]
> [bench.zip](https://github.com/potassco/clingoDL/files/4250511/bench.zip)
> 
> Added an ods file.
> clingodl-1 is the conda version, clingodl-3 is the wip which does not add satisfied clauses, and clingodl-2 is additionally sorting the edges.
> Not adding satisfied clauses actually has quite an effect (for what it does, or not does). Sorting additional has some influence and luck with some instances.

I have no clue what you have benchmarked. :smile: Can you also give the options?

**Comment by MaxOstrowski** [comment]
everything non clingo-dl should be hidden in the ods.No options used. Will work on a clean version...-------- Originalnachricht --------Betreff: Re: [potassco/clingoDL] sort edges during propagation to skip more edges (#25)Von: Roland Kaminski An: potassco/clingoDL Cc: Max Ostrowski ,Author 

bench.zip

Added an ods file.

clingodl-1 is the conda version, clingodl-3 is the wip which does not add satisfied clauses, and clingodl-2 is additionally sorting the edges.

Not adding satisfied clauses actually has quite an effect (for what it does, or not does). Sorting additional has some influence and luck with some instances.



I have no clue what you have benchmarked. 😄 Can you also give the options?



—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

[

{

"@context": "http://schema.org",

"@type": "EmailMessage",

"potentialAction": {

"@type": "ViewAction",

"target": "https://github.com/potassco/clingoDL/pull/25?email_source=notifications\u0026email_token=AEKNCPYSLDMO2LVRWDTVKI3REUW4XA5CNFSM4K3IO3Q2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM4I5XI#issuecomment-590909149",

"url": "https://github.com/potassco/clingoDL/pull/25?email_source=notifications\u0026email_token=AEKNCPYSLDMO2LVRWDTVKI3REUW4XA5CNFSM4K3IO3Q2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEM4I5XI#issuecomment-590909149",

"name": "View Pull Request"

},

"description": "View this Pull Request on GitHub",

"publisher": {

"@type": "Organization",

"name": "GitHub",

"url": "https://github.com"

}

}

]

**Comment by rkaminsk** [comment]
> everything non clingo-dl should be hidden in the ods. No options used. Will work on a clean version...

This is strange insofar that setting-1 and setting-3 should be equivalent then. `add_edge` should not add satisfied clauses if partial propagation is disabled. Why is setting-3 so much faster when there should be no difference at all?

**Comment by rkaminsk** [comment]
> everything non clingo-dl should be hidden in the ods. No options used. Will work on a clean version...

This is strange insofar that setting-1 and setting-3 should be equivalent then. `add_edge` should not add satisfied clauses if partial propagation is disabled. Why is setting-3 so much faster when there should be no difference at all?

**Comment by MaxOstrowski** [comment]
clingodl-1 was conda on potassco/label/dev a few weeks ago
I'm rerunning everything with a clean benchmark setup only using commit hashs for the different versions.

**Comment by rkaminsk** [comment]
> setting-1 was conda on potassco/label/dev a few weeks ago
> I'm rerunning everything with a clean benchmark setup only using commit hashs for the different versions.

I checked. There are no sat clauses if partial propagation is disabled. It would be a bug anyway and I don't think the algorithm could have worked correctly with it. It must be the removed overhead of solution checking.

**Comment by MaxOstrowski** [comment]
[clingodl.zip](https://github.com/potassco/clingoDL/files/4255718/clingodl.zip)

New results with hashes all look very very similar.
817cd87: master
b8ae258: wip
f4d5926: wip-1 (without satisfied clause removal)
33a94dc: feature/clauses (how is this different from wip ?
fe39c75: feature/skipedges (sorting)

**Comment by rkaminsk** [comment]
> [clingodl.zip](https://github.com/potassco/clingoDL/files/4255718/clingodl.zip)
> 
> New results with hashes all look very very similar.
> [817cd87](https://github.com/potassco/clingoDL/commit/817cd87e68a7a20667f19b50993ad4e891b0839a): master
> [b8ae258](https://github.com/potassco/clingoDL/commit/b8ae2589fed237518d19db161c4db808ed3678dd): wip
> [f4d5926](https://github.com/potassco/clingoDL/commit/f4d592698ba3c57c9b0fbc8e5a2203b99a0270b7): wip-1 (without satisfied clause removal)
> [33a94dc](https://github.com/potassco/clingoDL/commit/33a94dceb8046bfb40b4dea1161ed40949e01e00): feature/clauses (how is this different from wip ?
> [fe39c75](https://github.com/potassco/clingoDL/commit/fe39c75e25a816de0976beb494d1f96a4e2b4212): feature/skipedges (sorting)

I don't know what you are measuring but the conflict/choice counts are always the same (except for timeouts). I expect this to be the case for the first 4 settings because they are essentially the same. I also do not expect different runtimes except for noise. 

The last setting should result in different values. So I think there is a bug in you experiment.

Also please remove classes without choices/conflicts. It's just noise.

**Comment by MaxOstrowski** [comment]
[clingodl.zip](https://github.com/potassco/clingoDL/files/4261596/clingodl.zip)

Now we are talking.
The SAT test as well as the sorting makes a difference, even in timeouts!

**Comment by MaxOstrowski** [comment]
Going to make it right now, just some final questions:
> You should take a little more time with the code you are committing:
> 
>    - only add edges if they are active
>    - to avoid allocations, edges could be a member variable
>    - best make sorting an option right away so that benchmarking this gets easier
>    - I do not think that the cost of sorting will be a problem as the dijsktra algorithm is more expensive anyway.

1. Adding edges only if active could easily be done. Are you sure that "activeness" does not change during propagation ? Such that previous propagation could enable/disable this flag ?
2. edges could be a member, but then it would need to be a member of the state, and not the propagator, making it slightly more ugly ?
3. Could be done. Do you think this should be an option ? If sorting cost is negligible, having a heuristic instead of random order does not look like an option to me. Your choice.
4. Agreed, although it feels silly to sort static data every time :)

**Comment by rkaminsk** [comment]
> 1. Adding edges only if active could easily be done. Are you sure that "activeness" does not change during propagation ? Such that previous propagation could enable/disable this flag ?

Edges can only be disabled during propagation and only be re-enabled when undoing. With strong propagation it can in principle happen that a true edge is disabled. So to be on the safe side you would have to do the test twice. Btw. the other code you are executing out of order should not cause problems. But take that with a grain of salt I have not looked at the code in a long time. On a quick glance I noticed that the `false_lit_to_edges_` dictionary looks unnecessary. But I am for a complete rewrite anyway. The current code is ugly and one can see that it was written in haste. :wink:

> 2. edges could be a member, but then it would need to be a member of the state, and not the propagator, making it slightly more ugly ?

Not as ugly as other things in the code to squeeze out a bit of performance. In general, my rule of thumb is that if unnecessary allocations can be avoided and the code is performance critical, then avoid them. The vector should also not just be called edges but rather todo_edges_ or similar to indicate that it is something dynamic.

> 3. Could be done. Do you think this should be an option ? If sorting cost is negligible, having a heuristic instead of random order does not look like an option to me. Your choice.

But we do not even know if sorting is better than insertion order. And you already suggested additional sorting variants. I think an option is better.

> 4. Agreed, although it feels silly to sort static data every time :)

The problem is that a merge sort does not work in place. And even with a second vector it would perform a huge number of swaps. The standard quicksort does not have these problems. Another option would be to maintain the sorted slices in a tree like data structure. But I think this would just be unnecessarily complicated. Also in view of more dynamic heuristics, presorting is impossible.

**Comment by rkaminsk** [comment]
> [clingodl.zip](https://github.com/potassco/clingoDL/files/4261596/clingodl.zip)
> 
> Now we are talking.
> The SAT test as well as the sorting makes a difference, even in timeouts!

Nice, we should probably also test the remaining (partial) propagation modes also in combination with sorting. And maybe implement some different sorting heuristics. :wink:

**Comment by rkaminsk** [comment]
> [clingodl.zip](https://github.com/potassco/clingoDL/files/4261596/clingodl.zip)
> 
> Now we are talking.
> The SAT test as well as the sorting makes a difference, even in timeouts!

Nice, we should probably also test the remaining (partial) propagation modes also in combination with sorting. And maybe implement some different sorting heuristics. :wink:

**Comment by rkaminsk** [comment]
> [clingodl.zip](https://github.com/potassco/clingoDL/files/4261596/clingodl.zip)
> 
> Now we are talking.
> The SAT test as well as the sorting makes a difference, even in timeouts!

Nice, we should probably also test the remaining (partial) propagation modes also in combination with sorting. And maybe implement some different sorting heuristics. :wink:

**Comment by MaxOstrowski** [comment]
Comment Button not working, eh ? :laughing:

**Comment by rkaminsk** [comment]
The code is not indented properly and if you use todo_edges_ outside the state, then it should be called todo_edges. By convention a trailing underscore is used for private members.

The name "no-do-sort-edges" is horrible. The "do" is unnecessary. With the upcoming clingo-5.5, flags should not be started with "no-" because this would result in the options "--no-do-sort-edges" and "--no-no-do-sort-edges".

**Comment by MaxOstrowski** [comment]
fixed it and used --sort-edges.
The problem is, currently you can not disable this option, as there is no "no" option automatically generated.
In another issue on clingo(clingcon) you mentioned that clasp uses "--no-do-[option]".

**Comment by rkaminsk** [comment]
> The problem is, currently you can not disable this option, as there is no "no" option automatically generated.

Just update the submodule to the latest clingo wip then.

File: libclingo-dl/clingo-dl/propagator.hh [hunk 1]
```diff
@@ -175,6 +175,7 @@ struct PropagatorConfig {
     uint64_t mutex_cutoff{10};
     uint64_t propagate_root{0};
     uint64_t propagate_budget{0};
+    bool sort_edges{true};
```
Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh]: Please also consider the alignment.

```cpp
#include <iostream>

struct A {
    bool a;
    long b;
    bool c;
};
struct B {
    bool a;
    bool c;
    long b;
};
int main() {
    std::cerr << sizeof(A) << "vs." << sizeof(B) << std::endl;
}
```

Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh]: Do you want to make edge sorting the new default?

File: libclingo-dl/clingo-dl/propagator.hh [hunk 2]
```diff
@@ -1210,29 +1213,40 @@ public:
...
-                            state.stats.propagate_cost_add + state.propagate_budget > state.stats.propagate_cost_from + state.stats.propagate_cost_to);
-                    if (!propagate) { state.dl_graph.disable_propagate(); }
-                    // if !propgate -> can no longer propagate!
-                    if (propagate && !state.dl_graph.propagate(it->second, ctl)) { return; }
-                }
+                if (state.dl_graph.edge_is_active(it->second)) state.todo_edges.push_back(it->second);
+            }
+        }
+
+        if (!conf_.sort_edges) {
```
[the remark this answers is not in the document]
Comment (MaxOstrowski) [reply libclingo-dl/clingo-dl/propagator.hh]: thats a typo, thanks

File: libclingo-dl/clingo-dl/propagator.hh (hunk 1 above)
Comment (MaxOstrowski) [reply libclingo-dl/clingo-dl/propagator.hh]: As you like. I didn't thaught (at all. ) | (it matters for the Config).

Comment (MaxOstrowski) [reply libclingo-dl/clingo-dl/propagator.hh]: Actually yes, this is why we have a problem without the no-do-sort-thingy

Comment (rkaminsk) [reply libclingo-dl/clingo-dl/propagator.hh]: When writing C++ one has to think about these things.

Comment (rkaminsk) [reply libclingo-dl/clingo-dl/propagator.hh]: We don't. We are writing this code for the next clingo release. This is perfectly fine and I do it like this all the time. If you want to have this as the default I am fine with it.

Comment (MaxOstrowski) [reply libclingo-dl/clingo-dl/propagator.hh]: Great!

**Pull request merged**